### PR TITLE
WT-13445 Unit tests block_ext 4 - Insert functions with block

### PR DESCRIPTION
**Pull request by tod-johnson-mongodb**
[WT-13445](https://jira.mongodb.org/browse/WT-13445) Unit tests block_ext 4 - Insert functions with block
Unit tests for block manager block_ext.c part 4: extent insert/remove functions with a block parameter: __block_merge, __block_off_remove, __block_append, and __block_extend.

test/unittest/CMakeLists.txt: Add tests/block/test_extent_list_block.cpp
test/unittest/tests/block/test_extent_list_block.cpp: The new unit tests.

Error testing is included for only __block_extend. I wanted to get the normal tests done for all before adding error testing.

### CODE REVIEW

**Comment by tod-johnson-mongodb** [comment]
- The following patch build has succeeded so far but is still running: My Patch 298 https://spruce.mongodb.com/version/66d0d76f44f8460007d9452d/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC.
- The preceding patch build fully succeeded. It is the same except for comment and tags: My Patch 297 https://spruce.mongodb.com/version/66d0aa5360fef5000701173b/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

File: test/unittest/tests/block/test_extent_list_insert_block.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+TEST_CASE("Extent Lists: block_off_remove", "[extent_list]")
+{
+    /* Build Mock session, this will automatically create a mock connection. */
+    std::shared_ptr<MockSession> mock_session = MockSession::buildTestMockSession();
+    WT_SESSION_IMPL *session = mock_session->getWtSessionImpl();
+
+    std::vector<WT_EXT **> stack(WT_SKIP_MAXDEPTH, nullptr);
+
+    SECTION("remove multiple extents and verify all extents after each remove")
```
Comment (luke-pearson) [inline test/unittest/tests/block/test_extent_list_insert_block.cpp:143]: Is there any reason to define a `SECTION` if there's only one test in the overall test case?

File: test/unittest/tests/block/test_extent_list_insert_block.cpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        {
+            {off_size(4096, 2048), // First half of First [4,096, 6,143].
+              {
+                off_size(4096, 2048), // First [4,096, 6,143].
+              }},
+              {off_size(4096 + 2048, 2048), // Adjacent: Second half of First [6,144, 8,191].
+                {
+                  off_size(4096, 4096), // First [4,096, 8,191].
+                }},
+#if 0 // FAILED: REQUIRE( last == extlist.last ) with expansion: 0x0000000040319c10 ==.
```
Comment (luke-pearson) [inline test/unittest/tests/block/test_extent_list_insert_block.cpp:240]: Nice work finding this bug!

File: test/unittest/tests/block/test_extent_list_insert_block.cpp (hunk 1 above)
Comment (tod-johnson-mongodb) [reply test/unittest/tests/block/test_extent_list_insert_block.cpp:143]: Reasons to have a SECTION even if the TEST_CASE has only one SECTION:
* I imagine a muiti-level outline of tests: tag, file (maybe just tag is enough), TEST_CASE, SECTION, and test within the SECTION. The last level is needed if a SECTION contains multiple tests whether from a loop or otherwise.
* For consistency in how a TEST_CASE comment is written.  TEST_CASE gives the function, and SECTION describes what is tested. Without this a description of what is written might not be given.
* For ease of adding more SECTIONs to the TEST_CASE.
* If we eventually scan unit tests to find out what they test and construct a list for a test plan. No plans to do that.

File: test/unittest/tests/block/test_extent_list_insert_block.cpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+            {
+              off_size(4096, 4096),     // [4,096, 8,191] First.
+              off_size(5 * 4096, 4096), // [20,480, 24,575] Third.
+            }},
+          {4096, // [4,096, 8,191] First.
+            {
+              off_size(5 * 4096, 4096), // [20,480, 24,575] Third.
+            }},
+          {5 * 4096, // [20,480, 24,575] Third.
+            {}},
```
Comment (keitharnoldsmith) [inline test/unittest/tests/block/test_extent_list_insert_block.cpp:165]: Since there is an explicit test that the extent list is empty after performing all of the test cases, it might be worth a comment here (or when defining `test_list`, above, stating this assumption.

I.e., if somebody decides in the future to add/change the tests here, they might not realize the tests should leave the extent list empty.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM.

I've left a suggestion about adding a comment. But you don't need to have me re-review that.

File: test/unittest/tests/block/test_extent_list_insert_block.cpp (hunk 1 above)
Comment (luke-pearson) [reply test/unittest/tests/block/test_extent_list_insert_block.cpp:143]: Thanks Tod those reasons mostly make sense to me can you elaborate on this one?

> The last level is needed if a SECTION contains multiple tests whether from a loop or otherwise.

Why would the last level be needed if that singular section had multiple tests within it? It isn't like it would execute the reconstruction of the test much like other sections do.

**Review by luke-pearson — APPROVED** [review]
LGTM.

File: test/unittest/tests/block/test_extent_list_insert_block.cpp (hunk 1 above)
Comment (tod-johnson-mongodb) [reply test/unittest/tests/block/test_extent_list_insert_block.cpp:143]: The idea is to make a list of all tests that are run and what they do in sufficient detail so it is possible to check if anything is not tested that should be. With multiple subtests or loops there needs to be a statement of what each of them is the subtests or loop cycles is testing.  That is the lowest level.

File: test/unittest/tests/block/test_extent_list_insert_block.cpp (hunk 3 above)
Comment (tod-johnson-mongodb) [reply test/unittest/tests/block/test_extent_list_insert_block.cpp:165]: There is. nothing special about ending with an empty extent list. I just did that.

Every test has an expected result which depends upon both the test itself and every test that came before it. If any test is changed or added all the expected results after it must change, unless the added test is a NOP or the changed test has the same result as before.  That is implicit in testing the result after every test.

Comment (keitharnoldsmith) [reply test/unittest/tests/block/test_extent_list_insert_block.cpp:165]: Ah. I missed that. Ignore my suggestion then.
